File: src/main/java/edu/psu/ist/plato/kaiming/Machine.java
package edu.psu.ist.plato.kaiming;

import java.util.List;

import edu.psu.ist.plato.kaiming.arm64.ARM64Machine;
import edu.psu.ist.plato.kaiming.x86.X86Machine;

public abstract class Machine {
    public enum Arch {
        X86,
        ARM64,
    }
    
    public final Arch arch;
    
    protected Machine(Arch architecture) {
        arch = architecture;
    }
    
    public abstract List<MachRegister> registers();
    public abstract MachRegister returnRegister();
    public abstract int wordSizeInBits();
    public static final X86Machine x86 = X86Machine.instance;
    public static final ARM64Machine arm64 = ARM64Machine.instance;
}

File: src/main/scala/edu/psu/ist/plato/kaiming/Machine.scala
package edu.psu.ist.plato.kaiming

import ir._
import aarch64.AArch64Machine

abstract class Machine[A <: Arch] {

  val returnRegister: MachRegister[A]
  val wordSizeInBits: Int
  val registers: Set[MachRegister[A]]
  
  def toIRStatements(inst: Entry[A], builder: IRBuilder[A]): IRBuilder[A]

}

abstract class MachRegister[A <: Arch] {
  
  def name: String
  def sizeInBits: Int
  def containingRegister: MachRegister[A]
  def subsumedRegisters: Set[MachRegister[A]]
  
  override def equals(that: Any): Boolean
  override def hashCode: Int
  
}
